### PR TITLE
Add an ObjC class to act as the parent of all CF types.

### PR DESCRIPTION
**Pull request by DHowett-MSFT**
This pull request intends to both complete #719 and fix the crash behind #649 (#649 is both a compliance bug _and_ a crash bug.).

True Objective-C memory management for CoreFoundation (as seen [here](https://github.com/DHowett-MSFT/WinObjC/compare/201609-nscftype?expand=1)) is infeasible right now. We have too few fiddly bitfields and subclasses to determine with any accuracy whether an object was statically-allocated or heap-allocated, and we almost certainly don't want to manipulate the nonexistent reference count on non-heap-allocated objects.

Herein, therefore, we try to solve two of the major Objective-C <-> CF bridging compliance issues:
* Not every CF type has an Objective-C parallel (and as such cannot live in an Objective-C collection.)
* CF objects cannot be zombified.

### CODE REVIEW

**Comment by ms-jihua** [comment]
* Not every CF type has an Objective-C parallel (and as such cannot live in an Objective-C collection.)

Thank you for this this has been a thorn in my side for forever

File: Frameworks/CoreFoundation/Base.subproj/CFRuntime.c [hunk 1]
```diff
@@ -1118,6 +1128,11 @@ void __CFInitialize(void) {
         memset(__CFRuntimeClassTable, 0, sizeof(__CFRuntimeClassTable));
         memset(__CFRuntimeObjCClassTable, 0, sizeof(__CFRuntimeObjCClassTable));
 
+#if WINOBJC
```
Comment (ms-jihua) [inline Frameworks/CoreFoundation/Base.subproj/CFRuntime.c]: WINOBJC comment?

File: Frameworks/CoreFoundation/Base.subproj/CFRuntime.c [hunk 2]
```diff
@@ -1118,6 +1128,11 @@ void __CFInitialize(void) {
...
+        for (CFIndex idx = 1; idx < __CFRuntimeClassTableSize; ++idx) {
+            __CFRuntimeObjCClassTable[idx] = (uintptr_t)&_OBJC_CLASS__NSCFType;
+        }
```
Comment (ms-jihua) [inline Frameworks/CoreFoundation/Base.subproj/CFRuntime.c]: depending on how you feel about stl here, you could also use std::fill

**Comment by ms-jihua** [comment]
:shipit:

File: tests/unittests/Foundation/CoreFoundationTests.m [hunk 3]
```diff
@@ -56,3 +57,19 @@ static uint64_t walltime(const struct mach_timebase_info& timebase, uint64_t val
...
+    CFTreeContext treeContext{};
+    CFTypeRef tree = static_cast<CFTypeRef>(CFTreeCreate(kCFAllocatorDefault, &treeContext));
+
+    NSArray* array = nil;
+
+    // The ',' in this method call upsets EXPECT_NO_THROW(...), so we need to hide it behind a lambda.
+    auto create = [&tree](){
+        return [[NSArray alloc] initWithObjects:static_cast<id>(tree), nil];
+    };
+    EXPECT_NO_THROW(array = create());
```
Comment (rajsesh-msft) [inline tests/unittests/Foundation/CoreFoundationTests.m]: >EXPECT_NO_THROW [](start = 4, length = 15)

Not sure if these are ARM safe.

**Comment by rajsesh-msft** [comment]
:shipit:

**Comment by mnithish** [comment]
:shipit:

File: Frameworks/Foundation/NSObject.mm [hunk 4]
```diff
@@ -705,11 +708,36 @@ void WinObjC_SetMissingSelectorFatal(BOOL fatal) {
...
++ (BOOL)isSubclassOfClass:(Class)classRef {
+    Class curClass = self;
+
+    while (curClass != nil) {
+        if (curClass == classRef) {
+            return YES;
+        }
+
+        curClass = class_getSuperclass(curClass);
+    }
```
Comment (aballway) [inline Frameworks/Foundation/NSObject.mm]: would this work as a for loop?

**Comment by aballway** [comment]
:shipit:

File: Frameworks/Foundation/NSObject.mm (hunk 4 above)
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSObject.mm]: would this matter at all?

Comment (aballway) [inline Frameworks/Foundation/NSObject.mm]: nope, just think it looks a bit cleaner. Shoulda been a nit, my b

---
In reply to: [81378408](https://github.com/Microsoft/WinObjC/pull/1059#discussion_r81378408) [](ancestors = 81378408)

File: Frameworks/CoreFoundation/Base.subproj/CFRuntime.c (hunk 2 above)
Comment (DHowett-MSFT) [inline Frameworks/CoreFoundation/Base.subproj/CFRuntime.c]: I'd rather not get my peanut butter in my chocolate over here :)

---
In reply to: [81260899](https://github.com/Microsoft/WinObjC/pull/1059#discussion_r81260899) [](ancestors = 81260899)

File: tests/unittests/Foundation/CoreFoundationTests.m (hunk 3 above)
Comment (DHowett-MSFT) [inline tests/unittests/Foundation/CoreFoundationTests.m]: `EXPECT_*_THROW` are ARM-safe; it's `@catch(NSException*)` that's not ARM-safe :)

---
In reply to: [81263134](https://github.com/Microsoft/WinObjC/pull/1059#discussion_r81263134) [](ancestors = 81263134)

**Pull request merged**